Tuples: solve IDE crash with Extract Method

jcouv opened the pull request:
This is a possible design to implement the Extract Method refactoring on tuples. I'd like to get design feedback first, if it's holds water then go into finer-grain feedback.

Cyrus, I'd like to brainstorm with you on what tests I could add (I only included one simple case and did some manual checks).

Aleksey/Vlad, this makes use of `ITupleTypeSymbol` and adds some public APIs to the compiler. Is it reasonable?

@CyrusNajmabadi @VSadov @AlekseyTs for feedback.
CC @dotnet/roslyn-compiler @dotnet/roslyn-ide

Fixes #10570

jaredpar commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -2789,6 +2789,22 @@ protected override IPointerTypeSymbol CommonCreatePointerTypeSymbol(ITypeSymbol
+            {
+                return TupleTypeSymbol.Create(typesBuilder.ToImmutableAndFree(), default(ImmutableArray<string>), this);
+            }
+
```
Should we assert here that `elementTypes.Length == elementNames.Length`

jcouv replied:
I will throw with an error message if that's ok.

AlekseyTs commented on src/Compilers/Core/Portable/Symbols/ITupleTypeSymbol.cs:
```diff
@@ -14,7 +13,7 @@ namespace Microsoft.CodeAnalysis
     /// </remarks>
     public interface ITupleTypeSymbol : ITypeSymbol
```
I don't understand why this interface exists and who implements it.

AlekseyTs replied:
As I understand the plan was to delete this interface because we are not adding a special type kind for tuple type symbol. If we keep it, there should be a special type kind, special symbol kind and the symbol cannot be a named type symbol.

CyrusNajmabadi replied:
IDE would definitely like it.  Otherwise, it's going to be a PITA to work with a tuple.  If it's a named type instead, To find out the element types we'll have to look at .TypeArguments.  To find the tuple element names we'll have to look at all the fields.  etc. etc.  I personally like having this type.

CyrusNajmabadi replied:
Open questoins though: Should ITupleTypeSymbol be an INamedTypeSymbol?  I can see arguments both ways.

AlekseyTs replied:
We should simply extend INamedTypeSymbol for this. For example we didn't add new interface for anonymous types.

CyrusNajmabadi replied:
I'm ok with that.  *however* with the caveat that we add some methods (possibly extension) to INamedTypeSymbol to allow you to ask what the constituent set of tuple element types and tuple element names are.

CyrusNajmabadi replied:
NamedType is cleaner on the IDE for some things.

For example, we can then appropriately do cross language work more easily.  For example, if i generate method from C# to VB and i pass a tuple, then i can generate the right VBism properly.  And similarly, if i pass a wellformed tuple from VB to C# then that should also just work.  It makes a bit wonkier if the same metadata type appears to be different symbol types in VB and C#.

AlekseyTs replied:
When I say extend, I mean add these APIs to INamedTypeSymbol directly.

CyrusNajmabadi replied:
Great.

jcouv replied:
@AlekseyTs I was actually tempted to add a new `Kind` too, but it does seem this causes a lot of trouble. I'm ok to bring what I need into `INamedTypeSymbol` and remove `ITupleTypeSymbol`. This will connect well with what we discussed the other day too.

Should I just create three new members (one to check tuple type, one to get the types, one to get the members), or should I try to overload some existing members (for example `TypeArguments` could expose the types)?

jcouv replied:
@AlekseyTs Nevermind ;-) Reading the comments more carefully, I think you already clarified that new members would be better. I'll go that way.

AlekseyTs replied:
That is correct, new members are better.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -2789,6 +2789,22 @@ protected override IPointerTypeSymbol CommonCreatePointerTypeSymbol(ITypeSymbol
+            var typesBuilder = ArrayBuilder<TypeSymbol>.GetInstance(elementTypes.Length);
+            for (int i = 0; i < elementTypes.Length; i++)
+            {
+                typesBuilder.Add(elementTypes[i].EnsureCSharpSymbolOrNull<ITypeSymbol, TypeSymbol>(nameof(elementTypes)));
```
Feels like we should throw on null.

jcouv replied:
Yes, will do.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -92,6 +92,24 @@ DiagnosticBag diagnostics
+
+            if (numElements <= 1)
+            {
+                throw ExceptionUtilities.Unreachable;
```
Why is this unreachable?  It seems very possible that someone coudl pass in <= 1 elements.  If that's not allowed, we should throw a real argument exception.

jcouv replied:
Yes, I will add this check and throw corresponding exception in the entry point.

jcouv replied:
This and our discussion today at LDM (treating `ValueTuple` and `(tuple)` the same) makes me think that the code will need to accept 0-tuples and 1-tuples, even if we don't offer language sugar for it.  I'll discuss with the team.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/CodeActions/ExtractMethod/ExtractMethodTests.cs:
```diff
@@ -433,5 +433,14 @@ private static void NewMethod(out Construct obj1, out Construct obj2, out Constr
+        public async Task TestTuple()
+        {
+            await TestAsync(
+@"class Program { static void Main ( string [ ] args ) { [| (int, int) x = (1, 2); |]  System . Console . WriteLine ( x.Item1 ); } } ",
```
can you add a test with names on the left, names on the right, and names on both sides?

jcouv replied:
Will do. I'll also add a test with `var` (which I tested manually).

jcouv replied:
@CyrusNajmabadi Is there a reason for the weird spacing in those tests? I just mimicked other tests in the file...

CyrusNajmabadi replied:
We have a tool that generates these from within EditorTestApp.  This is how the tool formats things.

Note: you don't have tomimick the other tests.  You can write these tests over multiple lines, with whatever spacing you prefer if you want.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs:
```diff
@@ -145,6 +147,21 @@ public TypeSyntax CreateSimpleTypeSyntax(INamedTypeSymbol symbol)
+            private TupleTypeSyntax CreateTupleTypeSyntax(ITupleTypeSymbol symbol)
+            {
+                var list = new SeparatedSyntaxList<TupleElementSyntax>();
+                ImmutableArray<ITypeSymbol> types = symbol.Types;
```
var

jcouv replied:
Sure, I'll change.
FYI, on the compiler side, the guideline I heard were to spell out the type if it doesn't appear elsewhere on the line (explicit cast, constructor, etc).

DustinCampbell replied:
Yup -- much of the code base outside of the compiler, including workspaces and IDE, uses var. Different teams. Different religions. :smile:

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs:
```diff
@@ -145,6 +147,21 @@ public TypeSyntax CreateSimpleTypeSyntax(INamedTypeSymbol symbol)
+            {
+                var list = new SeparatedSyntaxList<TupleElementSyntax>();
+                ImmutableArray<ITypeSymbol> types = symbol.Types;
+                ImmutableArray<string> names = symbol.Names;
```
var.

jcouv replied:
Done.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs:
```diff
@@ -145,6 +147,21 @@ public TypeSyntax CreateSimpleTypeSyntax(INamedTypeSymbol symbol)
+                bool hasNames = !names.IsDefault;
+                for (int i = 0; i < types.Length; i++)
+                {
+                    IdentifierNameSyntax name = hasNames ? SyntaxFactory.IdentifierName(names[i]) : default(IdentifierNameSyntax);
```
var

CyrusNajmabadi replied:
"null" instead of default(IdentifierName)

jcouv replied:
Done (both).

CyrusNajmabadi commented on src/Workspaces/Core/Portable/SymbolId/SymbolKey.TupleTypeSymbolKey.cs:
```diff
@@ -0,0 +1,60 @@
+            {
+                _types = symbol.Types.Select(t => GetOrCreate(t, visitor)).ToArray();
+
+                ImmutableArray<string> names = symbol.Names;
```
inline this?

jcouv replied:
Done.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/SymbolId/SymbolKey.TupleTypeSymbolKey.cs:
```diff
@@ -0,0 +1,60 @@
+
+            public override SymbolKeyResolution Resolve(Compilation compilation, bool ignoreAssemblyKey, CancellationToken cancellationToken)
+            {
+                IEnumerable<ITupleTypeSymbol> tupleType = Resolve(compilation, ignoreAssemblyKey);
```
inline this?

jcouv replied:
Done.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/SymbolId/SymbolKey.TupleTypeSymbolKey.cs:
```diff
@@ -0,0 +1,60 @@
+            {
+                ITypeSymbol[] types = _types.Select(a => a.Resolve(compilation, ignoreAssemblyKey).Symbol as ITypeSymbol).ToArray();
+                if (types.Any(a => a == null))
+                {
```
comment this (like we do in named type resolyution).

jcouv replied:
Sure.
However, I didn't find which comment you're referring to. I'll craft one.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/SymbolId/SymbolKey.TupleTypeSymbolKey.cs:
```diff
@@ -0,0 +1,60 @@
+
+                return SequenceEquals(other._types, _types, comparer)
+                       && ((other._names == null && _names == null)
+                           || SequenceEquals(other._names, _names, CaseInsensitiveComparison.Comparer));
```
SequenceEquals handles the null case.

jcouv replied:
Oops. I had checked but misread the code. The case of two nulls is covered in the first check...

CyrusNajmabadi commented on src/Workspaces/Core/Portable/SymbolId/SymbolKey.TupleTypeSymbolKey.cs:
```diff
@@ -0,0 +1,60 @@
+
+            internal override int GetHashCode(ComparisonOptions options)
+            {
+                return Hash.CombineValues(_types);
```
Ignoring names intentionally?  If so, put a comment explaining why.

jcouv replied:
Yes that's intentional. I think types are good enough for hash code. I'll add comment.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/SymbolId/SymbolKey.Visitor.cs:
```diff
@@ -71,9 +67,17 @@ public override SymbolKey VisitModule(IModuleSymbol moduleSymbol)
+
+                if (namedTypeSymbol.IsTupleType)
+                {
+                    return new TupleTypeSymbolKey((ITupleTypeSymbol)namedTypeSymbol, this);
```
ITupleTypeSymbol didn't derive from NamedTypeSymbol.  So it's super weird that this works.

Also, shouldn't we have VisitTupleType, instead of VisitNamedType?

jcouv replied:
Right now, `TupleTypeSymbol` derives from `NamedTypeSymbol`.
I'll remove `ITupleTypeSymbol` as Aleksey suggested.

jcouv commented:
@AlekseyTs @CyrusNajmabadi @jaredpar 
The latest commit addresses the initial feedback.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -2789,6 +2789,37 @@ protected override IPointerTypeSymbol CommonCreatePointerTypeSymbol(ITypeSymbol
             return CreatePointerTypeSymbol(elementType.EnsureCSharpSymbolOrNull<ITypeSymbol, TypeSymbol>("elementType"));
         }
 
+        protected override INamedTypeSymbol CommonCreateTupleTypeSymbol(ITypeSymbol[] elementTypes, string[] elementNames)
```
Why not immutable arrays? Is there a pattern we are following?

AlekseyTs replied:
I do not see a single test for this function. At the very least every branch should be covered with a test.

jcouv replied:
I had looked at other methods in the same class but didn't find a clear standard. But I agree that would be a good thing to have.
Most methods seem to use `IEnumerable`, some use `ImmutableArray`, `ICollection` and a couple use `ArrayBuilder`. I thought some also used arrays, but I was mistaken (they are only `params`).
I'll change to `ImmutableArray`.

I'll also add direct tests for this new API.

CyrusNajmabadi replied:
The closest analog to this would be INamedTypeSymbol.Construct.  So follow whatever that does.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -2789,6 +2789,37 @@ protected override IPointerTypeSymbol CommonCreatePointerTypeSymbol(ITypeSymbol
 
+        protected override INamedTypeSymbol CommonCreateTupleTypeSymbol(ITypeSymbol[] elementTypes, string[] elementNames)
+        {
+            if (elementTypes.Length <= 1)
```
Should we check for null?

jcouv replied:
Yes. Done.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -2789,6 +2789,37 @@ protected override IPointerTypeSymbol CommonCreatePointerTypeSymbol(ITypeSymbol
+                return TupleTypeSymbol.Create(typesBuilder.ToImmutableAndFree(), default(ImmutableArray<string>), this);
+            }
+
+            return TupleTypeSymbol.Create(typesBuilder.ToImmutableAndFree(), elementNames.ToImmutableArrayOrEmpty(), this);
```
Shouldn't null elementNames be translated to default immutable array?

jcouv replied:
After changing `elementNames` to `ImmutableArray<string>`, I can pass it directly into `TupleTypeSymbol` whether it have values or it is default.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -92,6 +92,24 @@ DiagnosticBag diagnostics
+            {
+                throw ExceptionUtilities.Unreachable;
+            }
+            NamedTypeSymbol underlyingType = GetTupleUnderlyingType(elementTypes, compilation);
```
Consider adding an empty line after the 'if'

jcouv replied:
Done.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -269,6 +287,37 @@ internal static int NumberOfValueTuples(int numElements, out int remainder)
         /// <summary>
         /// Produces the underlying ValueTuple corresponding to this list of element types.
         /// </summary>
+        private static NamedTypeSymbol GetTupleUnderlyingType(ImmutableArray<TypeSymbol> elementTypes, CSharpCompilation compilation)
```
It feels like there is a lot of code is getting duplicated here. The function below can be changed to 
- take a compilation instead of binder, logic in Binder.GetWellKnownType isn't that complex to duplicate, just a Binder.ReportUseSiteDiagnostics call after Compilation.GetWellKnownType;
- report diagnostics only if bag is not null.

jcouv replied:
I was not too happy about this duplication either. Thanks for the tips for how to do it.
One other trouble I have here is that the original method needed a syntax node (to get a location in reported diagnostics), but in the new scenario I have no node.

AlekseyTs replied:
And no need to report diagnostics.

AlekseyTs replied:
null should work just fine.

jcouv replied:
Done.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -514,6 +563,27 @@ internal NamedTypeSymbol UnderlyingTupleType
+        {
+            get
+            {
+                return _fields.Select(f => (ITypeSymbol)f.Type).ToImmutableArray();
```
I believe there is a SelectAsArray helper. Or something similar.

jcouv replied:
Found it. Thanks

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -514,6 +563,27 @@ internal NamedTypeSymbol UnderlyingTupleType
+            {
+                if (_hasFriendlyNames)
+                {
+                    return _fields.Select(f => f.Name).ToImmutableArray();
```
Same here.

jcouv replied:
Done.

AlekseyTs commented:
I am done with review pass for changes under src/Compilers

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -823,6 +823,19 @@ public INamedTypeSymbol GetTypeByMetadataName(string fullyQualifiedMetadataName)
+        /// <remarks>
+        /// Since VB doesn't support tuples yet, this call will fail in a VB compilation.
+        /// </remarks>
+        public INamedTypeSymbol CreateTupleTypeSymbol(ITypeSymbol[] elementTypes, string[] elementNames = null)
```
@CyrusNajmabadi How did you manage without similar method for anonymous types?

CyrusNajmabadi replied:
Anonymous types can't cross methods.  So we never needed to be able to construct symbols for them to put in signatures in a cross language scenario.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb:
```diff
@@ -1201,6 +1201,18 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
 
+        Public ReadOnly Property TupleElementTypes As ImmutableArray(Of ITypeSymbol) Implements INamedTypeSymbol.TupleElementTypes
+            Get
+                Throw New NotImplementedException() ' PROTOTYPE(tuples)
```
Not appropriate to throw. VB doesn't support tuples.

jcouv replied:
Replaced with a `NotSupportedException`

jcouv replied:
Changed to return nothing, which will match C#.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb:
```diff
@@ -1201,6 +1201,18 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+
+        Public ReadOnly Property TupleElementNames As ImmutableArray(Of String) Implements INamedTypeSymbol.TupleElementNames
+            Get
+                Throw New NotImplementedException() ' PROTOTYPE(tuples)
```
Not appropriate to throw. VB doesn't support tuples.

jcouv replied:
Also replaced with a `NotSupportedException`

jcouv replied:
Changed to return nothing, which will match C#.

jaredpar commented on src/Workspaces/Core/Portable/SymbolId/SymbolKey.TupleTypeSymbolKey.cs:
```diff
@@ -0,0 +1,61 @@
+                var comparer = SymbolKeyComparer.GetComparer(options);
+
+                return SequenceEquals(other._types, _types, comparer)
+                       && SequenceEquals(other._names, _names, CaseInsensitiveComparison.Comparer);
```
Why use the insensitive comparison here?  I'm not entirely familiar with this type but names in C# are case sensitive.

jcouv replied:
That's a mistake. I'll fix.

jaredpar commented on src/Workspaces/Core/Portable/SymbolId/SymbolKey.TupleTypeSymbolKey.cs:
```diff
@@ -0,0 +1,61 @@
+                    yield break;
+                }
+
+                yield return compilation.CreateTupleTypeSymbol(types, _names);
```
Should we use an singelton enumerable here vs. an iterator?  Iterator seems like a large overhead for this scenario.

CyrusNajmabadi replied:
I'm not too concerned.  SymbolKey resolution is not really a hot path.  Generally what happens is we make some symbol keys for a handful of symbols in one compilation, fork the compilation, and then try to resolve tehm in the new compilation.

jaredpar replied:
True.  But doing it right takes less keyboard time than this conversation ;)

VSadov replied:
returning single or empty enumerables would also be more readable in this case. IMO.

jcouv replied:
Fixed with returning empty or singleton from `SpecializedCollections`.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -2789,11 +2789,26 @@ protected override IPointerTypeSymbol CommonCreatePointerTypeSymbol(ITypeSymbol
             {
+                if (elementTypes[i] == null)
+                {
+                    throw new ArgumentNullException($"{nameof(elementTypes)}[{i}]");
```
cute.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs:
```diff
@@ -556,13 +556,17 @@ public virtual bool IsAnonymousType
+        /// <summary>
+        /// If this symbol represents a tuple type, get the types of the tuple's elements.
+        /// </summary>
+        public virtual ImmutableArray<ITypeSymbol> TupleElementTypes => default(ImmutableArray<ITypeSymbol>);
```
Consider:  throw in the non-IsTupleType case.

jcouv replied:
From discussion with Aleksey, I will change the VB side to match this (return nothing) to align.

CyrusNajmabadi commented on src/Compilers/Core/Portable/CodeAnalysisResources.resx:
```diff
@@ -510,4 +510,10 @@
 {0}</value>
   </data>
+  <data name="TupleNamesAllOrNone" xml:space="preserve">
+    <value>If names used in a tuple type, then there must be a name for every member.</value>
```
If names *are* used

jcouv replied:
Thanks. Fixed.

CyrusNajmabadi commented on src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb:
```diff
@@ -2597,8 +2597,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-        Protected Overrides Function CommonCreateTupleTypeSymbol(elementTypes As ITypeSymbol(), elementNames As String()) As ITupleTypeSymbol
-            Throw New NotSupportedException(VBResources.ThereAreNoPointerTypesInVB)
+        Protected Overrides Function CommonCreateTupleTypeSymbol(elementTypes As ITypeSymbol(), elementNames As String()) As INamedTypeSymbol
+            Throw New NotSupportedException(VBResources.ThereAreNoPointerTypesInVB) ' PROTOTYPE(tuples)
```
Wrong resource.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/SymbolId/SymbolKey.TupleTypeSymbolKey.cs:
```diff
@@ -47,12 +48,12 @@ internal override bool Equals(TupleTypeSymbolKey other, ComparisonOptions option
                 return SequenceEquals(other._types, _types, comparer)
-                       && ((other._names == null && _names == null)
-                           || SequenceEquals(other._names, _names, CaseInsensitiveComparison.Comparer));
+                       && SequenceEquals(other._names, _names, CaseInsensitiveComparison.Comparer);
```
Is CaseInsensitive appropriate here?  What do we do elsewhere in symbol key resolution when we're looking at names?

jcouv replied:
Yes, that's a bug. Fixed.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1,5 +1,15 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using Microsoft.CodeAnalysis.CodeGen;
```
Fix your VS settings.  'System' namepsaces should come first.

jcouv replied:
That explains why I've been messing up usings... :-(
Thanks

leppie replied:
@jcouv: VS 'remote and sort' will do that automatically

jcouv replied:
@leppie I usually invoke "Sort usings" only. But my VS setting didn't prioritize "System" namespaces first.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/CodeActions/ExtractMethod/ExtractMethodTests.cs:
```diff
@@ -433,5 +433,50 @@ private static void NewMethod(out Construct obj1, out Construct obj2, out Constr
+        {
+            await TestAsync(
+@"class Program { static void Main ( string [ ] args ) { [| (int a, int b) x = (c: 1, d: 2); |]  System . Console . WriteLine ( x.a ); } } ",
+@"class Program { static void Main ( string [ ] args ) { (int a, int b) x = {|Rename:NewMethod|}(); System.Console.WriteLine(x.a); } private static (int a, int b) NewMethod() { return (c: 1, d: 2); } }",
```
@DustinCampbell This is interesting.  because tuple types are so compatible, it's acceptable for NewMethod to return ```(int a, int b)``` or ```(int c, int d)```.  I'm curious what we think customers would prefer in a case like this.

Personally, i'm not sure.

VSadov replied:
return (c: 1, d: 2);    . is least  lossy so it might be mildly preferable.

Note - you can also subsequently simplify the return to just    "return    (1, 2)"  
Unless you are in a lambda, where you might be inferring the return type from return values.

CyrusNajmabadi commented:
outside ofa  few nits the IDE side LGTM.

jcouv commented:
@CyrusNajmabadi @AlekseyTs @jaredpar 
Newest commit addresses the second round of PR feedback.

CyrusNajmabadi commented:
IDE LGTM

VSadov commented:
LGTM

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -268,21 +269,45 @@ internal static int NumberOfValueTuples(int numElements, out int remainder)
+                Binder.ReportUseSiteDiagnostics(firstTupleType, diagnostics, syntax);
+            }
+
+            if ((object)firstTupleType == null)
```
This can never happen.

jcouv replied:
Thanks. Fixed

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -268,21 +269,45 @@ internal static int NumberOfValueTuples(int numElements, out int remainder)
         /// Produces the underlying ValueTuple corresponding to this list of element types.
+        ///
+        /// Pass a null diagnostic bag and syntax node if you don't care about diagnostics.
+        /// May return null.
```
This should never happen. It may return an error type symbol though.

jcouv replied:
Removed.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -268,21 +269,45 @@ internal static int NumberOfValueTuples(int numElements, out int remainder)
+                    Binder.ReportUseSiteDiagnostics(firstTupleType, diagnostics, syntax);
+                }
+
+                if ((object)chainedTupleType == null)
```
This can never happen

jcouv replied:
Removed null check.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -514,6 +539,27 @@ internal NamedTypeSymbol UnderlyingTupleType
 
         internal bool HasFriendlyNames => _hasFriendlyNames;
 
+        public override ImmutableArray<ITypeSymbol> TupleElementTypes
```
Consider making this an explicit implementation. We don't want to use this method within a compiler by accident.

AlekseyTs replied:
Scratch that.

AlekseyTs commented:
Still don't see any tests in compilers.

jcouv commented:
Sorry, I forgot to add the tests for new public API. I'll add.

jcouv commented:
@AlekseyTs The newest commit adds the tests for new API on Compilation.
I also removed a couple of references to `ITupleTypeSymbol` that were left and added other refactoring tests that I was working on.

jaredpar commented on src/Compilers/CSharp/Test/Symbol/Compilation/CompilationAPITests.cs:
```diff
@@ -2053,6 +2053,30 @@ public void PreviousSubmissionWithError()
         }
 
+        [Fact]
+        public void CreateTupleTypeSymbol()
```
Have we annotated this **class** with `[CompilerTrait(CompilerFeature.Tuples)]`? If not we should to facilitate tracking these tests.  I'm really curious how many we have for tuples.  Interested in comparing it to say local functions.

jaredpar replied:
Consider the case we are using a `Compilation` that doesn't include a valid definition for `ValueTuple` (for example compiling against vanilla 4.6).  How should this API behave in that case?  I think we should add a test for that.

AlekseyTs replied:
I think it would be inappropriate to annotate the class because it is not specific to tuples. Perhaps the individual tests should be annotated instead. I personally would keep all tuple specific test in one file rather than spreading them around.

AlekseyTs replied:
I meant compiler tests, of course.

jaredpar replied:
@AlekseyTs good catch. I mistakenly thought this was in a type that contained only tuple tests.

jcouv replied:
Sure, I'll move the test along with most tuples tests in `CodeGenTupleTests.cs` for the C# side. I will leave the VB ones as-is.

jcouv replied:
@jaredpar, I don't mind annotating the tests (although I could not find `CompilerTrait` or `CompilerFeature`). But I don't much care about test counts as a measure.

One more thought on test organization, the current tests seem to be structured (parsing, symbol, semantic, emit). I don't mind moving away from that model, and towards one-test-file-per-new-feature, but it would be good to make it an explicit guideline.

jcouv replied:
@jaredpar Is this what you meant: `[Trait(Traits.Feature, Traits.Features.Tuples)]`?

jaredpar replied:
@jcouv it's possible the change to add `CompilerTrait` hasn't yet sync'd into the tuple branch.  If it's not resolvable please add a work item to add it back later once the branches merge. 

>  But I don't much care about test counts as a measure.

It provides value in the following two ways:

1. Roslyn has ~65,000 tests layered and ordered in no standard way.  This makes it impossible for people, aka leads, to ask questions like "what all tests do we have against tuples, or what tests have tuples and dynamic"?  Having this data makes it possible to audit test sets to determine if there are gaps that need to be plugged.  
2. It's by no means "the" metric for quality but it is "a" metric.  It's useful to comparce similar features for test size to see if there is a gap worth looking into.  If Local functions have 50 tests and tuples 500 then we likely have an issue.

jcouv replied:
The case where `Compilation` doesn't contain `ValueTuple` is what is tested here. Your question makes me think that the tuple should be error type if the underlying type is error. I will change and test that. 
I think this also goes in the direction of unifying tuple and underlying type.

jaredpar commented on src/Compilers/CSharp/Test/Symbol/Compilation/CompilationAPITests.cs:
```diff
@@ -2053,6 +2053,30 @@ public void PreviousSubmissionWithError()
+            Assert.Throws<ArgumentException>(() => comp.CreateTupleTypeSymbol(ImmutableArray<ITypeSymbol>.Empty, default(ImmutableArray<string>)));
+            Assert.Throws<ArgumentException>(() => comp.CreateTupleTypeSymbol(new[] { intType }.AsImmutable(), default(ImmutableArray<string>)));
+
+            Assert.Throws<ArgumentException>(() => comp.CreateTupleTypeSymbol(new[] { intType, intType }.AsImmutable(), new[] { "Item1" }.AsImmutable()));
```
What about the compiler scenario where we use reserved names like Item1?  This test doesn't seem to cover that, or at least it's misleading, because it's hitting a different case: tuple type count and element count differ.

jaredpar replied:
Seems like there are a number of other cases that would be interesting to test here:

- Names are illegal C# identifiers. 
- Names are blank.
- Names are non-reserved identifiers and differ in count than the type list.

AlekseyTs replied:
Consider scenario when VB symbol is passed for a type.

jcouv replied:
@jaredpar This new API is a veneer on existing underlying `TupleTypeSymbol` construction. Using `Item1` in second position is already covered for instance. I'll check whether we already cover the scenarios you listed (I don't think we check illegal C# identifiers for instance).
I don't see much advantage to duplicate coverage though. Do you prefer to test such cases through this new `Compilation.CreateTupleTypeSymbol` or through `Compilation.Compile` (as many of our existing tests do)?

jcouv replied:
@AlekseyTs Sure. I may have a question on how to do that, but I'll see if I can find an example first.

jaredpar commented on src/Compilers/CSharp/Test/Symbol/Compilation/CompilationAPITests.cs:
```diff
@@ -2053,6 +2053,30 @@ public void PreviousSubmissionWithError()
+            Assert.Throws<ArgumentNullException>(() => comp.CreateTupleTypeSymbol(default(ImmutableArray<ITypeSymbol>), default(ImmutableArray<string>)));
+
+            Assert.Throws<ArgumentException>(() => comp.CreateTupleTypeSymbol(ImmutableArray<ITypeSymbol>.Empty, default(ImmutableArray<string>)));
+            Assert.Throws<ArgumentException>(() => comp.CreateTupleTypeSymbol(new[] { intType }.AsImmutable(), default(ImmutableArray<string>)));
```
Unclear why this is illegal.  Is it because it's a one element tuple?  May make sense to have this as a separate test or commented so the intent is clearer.

AlekseyTs replied:
It would be good to have a success test for two types and default names.

jaredpar replied:
We should likely also have a test for the case of 8 and 9 element tuples.  They are interesting cases for the language hence it reasons they should be interesting for the API as well.

AlekseyTs replied:
I see this scenario covered below.

jaredpar replied:
@AlekseyTs our comments crossed and it's unclear if are you referring to "two types and default names" or "8 and 9 element tuples"?  I believe I see the former on line 2071.  I don't see the latter (maybe it's in another file)?

AlekseyTs replied:
Sorry, I was referring to my comment about two types ...

jcouv replied:
My focus here is to test the new code. I'm hesitant to recreate a new matrix for every entry point that hits the core logic of the symbol, but I'm open a reasonable subset. 
I'll split the test to make it easier to follow, add some success cases, add VB types, and add long tuples.

jcouv replied:
On second thought, this new path bypasses name validation, so it is worthwhile to add :-S

jcouv replied:
Aleksey clarified that current behavior is indeed correct. `CreateTupleTypeSymbol` is a low-level API which doesn't offer diagnostics and bad names should only be checked when constructed from source. 
I will leave code as-is, but add tests to make this behavior explicit.

jaredpar commented on src/Compilers/CSharp/Test/Symbol/Compilation/CompilationAPITests.cs:
```diff
@@ -2053,6 +2053,30 @@ public void PreviousSubmissionWithError()
+
+            var tupleWithNames = comp.CreateTupleTypeSymbol(new[] { intType, stringType }.AsImmutable(), new[] { "Alice", "Bob" }.AsImmutable());
+            Assert.True(tupleWithNames.IsTupleType);
+            Assert.Equal("(System.Int32 Alice, System.String Bob)", tupleWithNames.ToTestDisplayString());
```
Do we have tests to validate the actual name elemens directly return Alice and Bob?

jcouv replied:
Yes, although most coverage is indirect.
For instance, in `LongTupleDeclarationWithNames` we execute and verify:
```C#
class C
{
    static void Main()
    {
        (int a, int b, int c, int d, int e, int f, int g, string h, int i, int j, int k, int l) x = (1, 2, 3, 4, 5, 6, 7, ""Alice"", 2, 3, 4, 5);
        System.Console.WriteLine($""{x.a} {x.b} {x.c} {x.d} {x.e} {x.f} {x.g} {x.h} {x.i} {x.j} {x.k} {x.l}"");
    }
}
```

jcouv replied:
I'll add explicit test.

jaredpar commented on src/Compilers/CSharp/Test/Symbol/Compilation/CompilationAPITests.cs:
```diff
@@ -2053,6 +2053,30 @@ public void PreviousSubmissionWithError()
+            Assert.Throws<ArgumentNullException>(() => comp.CreateTupleTypeSymbol(new[] { intType, null }.AsImmutable(), default(ImmutableArray<string>)));
+
+            var tupleWithoutNames = comp.CreateTupleTypeSymbol(new[] { intType, stringType }.AsImmutable(), default(ImmutableArray<string>));
+            Assert.True(tupleWithoutNames.IsTupleType);
```
What happens if the type I pass as an element type is an error type?  What will the properties of the resulting tuple be?

jcouv replied:
That's a good question. Current behavior as I understand it is nothing special will happen. You will just get a tuple with an error type. 
Should an error type somehow taint the tuple itself?

jaredpar replied:
@jcouv it seems like a reasonable behavior.  But I think we should document the expectation.

jcouv replied:
From discussion with Aleksey, we clarified the current behavior regarding error type is acceptable. The symbol is good even if a part is an error type (that is similar to array and other symbols).

jaredpar commented on src/Compilers/CSharp/Test/Symbol/Compilation/CompilationAPITests.cs:
```diff
@@ -2053,6 +2053,30 @@ public void PreviousSubmissionWithError()
+            Assert.True(tupleWithoutNames.IsTupleType);
+            Assert.Equal("(System.Int32, System.String)", tupleWithoutNames.ToTestDisplayString());
+
+            var tupleWithNames = comp.CreateTupleTypeSymbol(new[] { intType, stringType }.AsImmutable(), new[] { "Alice", "Bob" }.AsImmutable());
```
FYI: `ImmutableArray.Create` is another option here.  Not being prescriptive here.

jcouv replied:
Good to know. Thanks

jaredpar commented on src/Compilers/Core/Portable/Symbols/INamedTypeSymbol.cs:
```diff
@@ -134,5 +134,8 @@ public interface INamedTypeSymbol : ITypeSymbol
         bool MightContainExtensionMethods { get; }
+
+        ImmutableArray<ITypeSymbol> TupleElementTypes { get; }
+        ImmutableArray<string> TupleElementNames { get; }
```
What happens when I call these methods on non-tuple types?  For instance what happens when this is invoked on say `System.String`?  

Do we have tests for that case?

jaredpar replied:
Given this is a shared symbol probably need same tests for both languages.

jcouv replied:
We already have test for these methods in VB (always returns nothing) and positive case in C# (returns proper arrays on tuple types). 
I'll add negative case for C# (returns default arrays on non-tuple types).

jcouv replied:
Added `TupleMethodsOnNonTupleType` in upcoming commit.

jaredpar commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -514,6 +528,27 @@ internal NamedTypeSymbol UnderlyingTupleType
+        {
+            get
+            {
+                return _fields.SelectAsArray(f => (ITypeSymbol)f.Type);
```
On non-tuple types `TupleElementNames` will return an `ImmutableArray<string>` that has the value `default(ImmutableArray<string>)`.  This method returns `ImmutableArray<ITypeSymbol>.Empty`.  These are semantically different values (uninitialized array vs. empty array).  It seems incorrect to differ here.

AlekseyTs replied:
I don't think this method ever returns an empty array, there are always fields in a tuple type.

jcouv replied:
Yes, a tuple always has types (even a theoretical 0-tuple would have an empty list of tuples), but it may have no friendly names (as in default/null). I think the current behavior is correct.

jaredpar commented on src/Compilers/VisualBasic/Test/Semantic/Compilation/CompilationAPITests.vb:
```diff
@@ -1206,6 +1206,12 @@ BC2014: the value '_' is invalid for option 'RootNamespace'
         End Sub
 
         <Fact()>
+        Public Sub TuplesNotSupported()
```
Should add `<CompilerTrait(CompilerFeature.Tuples)>`

jcouv replied:
Done.

jaredpar commented on src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/FieldTests.vb:
```diff
@@ -45,6 +45,25 @@ End Structure
             CompilationUtils.AssertNoDeclarationDiagnostics(compilation)
         End Sub
 
+        <Fact>
```
Should add `<CompilerTrait(CompilerFeature.Tuples)>`

jcouv replied:
Done.

jaredpar commented on src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs:
```diff
@@ -145,6 +147,22 @@ public TypeSyntax CreateSimpleTypeSyntax(INamedTypeSymbol symbol)
+                var list = new SeparatedSyntaxList<TupleElementSyntax>();
+                var types = symbol.TupleElementTypes;
+                var names = symbol.TupleElementNames;
+                bool hasNames = !names.IsDefault;
```
This check here seems to indicate the contract for `TupleElementNames` is to return `default(ImmutableArray<string>)` in the case of no names vs. `ImmutableArray<string>.Empty`.  We shuold make sure there is a test for this.

jcouv replied:
Yes, added.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/FieldTests.vb:
```diff
@@ -45,6 +45,25 @@ End Structure
+            Dim fieldType = DirectCast(field.Type, INamedTypeSymbol)
+
+            Assert.False(fieldType.IsTupleType)
+            Assert.Equal(Nothing, fieldType.TupleElementTypes)
```
I have no idea what XUnit does in this case. Consider verifying that the array .IsDefault is true instead.

jcouv replied:
Done.

AlekseyTs commented on src/EditorFeatures/CSharpTest/CodeActions/EncapsulateField/EncapsulateFieldTests.cs:
```diff
@@ -1255,5 +1255,89 @@ public int I
+
+    void baz()
+    {
+        var q = bar;
```
As @jaredpar mentioned on some other PR, words like foo, bar, ... are getting flagged by a polycheck tool. It is better to not use them at all rather than confirming that context is appropriate for all occurrences.

jcouv replied:
Sure, I'll change. Should the tests above that use the same pattern be updated too?

jcouv commented:
Latest commit adds the suggested tests and fixes a bug with checking null tuple element names.

The tuple tests are marked similarly to other feature tests, but I filled issue #10794 to upgrade the trait once we integrate `CompilerTrait`.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:
```diff
@@ -2192,6 +2195,242 @@ static void Main()
+            Assert.False(((TupleTypeSymbol)tupleWithoutNames).HasFriendlyNames);
+            Assert.True(tupleWithoutNames.TupleElementNames.IsDefault);
+            Assert.Equal(new[] { "System.Int32", "System.String" }, tupleWithoutNames.TupleElementTypes.Select(t => t.ToTestDisplayString()));
+            Assert.Equal(SymbolKind.NamedType, tupleWithoutNames.Kind);
```
I would expect the same asserts in the CreateTupleTypeSymbol_WithValueTuple unit-test. Presence or lack of the underlying type shouldn't have any effect on the new APIs.

jcouv replied:
Done.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:
```diff
@@ -2192,6 +2195,242 @@ static void Main()
+            var comp = CSharpCompilation.Create("test", references: new[] { MscorlibRef, tupleComp.ToMetadataReference() });
+
+            ITypeSymbol intType = comp.GetSpecialType(SpecialType.System_Int32);
+            ITypeSymbol errorType = comp.GetWellKnownType(WellKnownType.System_Runtime_CompilerServices_DynamicAttribute);
```
I think you can simply use ErrorTypeSymbol.UnknownResultType.

jcouv replied:
Done.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:
```diff
@@ -2192,6 +2195,242 @@ static void Main()
+            var comp = CreateCSharpCompilation("test", "", referencedCompilations: new[] { vbComp }); // no ValueTuple
+
+            ITypeSymbol vbType = (ITypeSymbol)comp.GetMember("C");
+            var tupleWithNames = comp.CreateTupleTypeSymbol(ImmutableArray.Create(vbType, vbType), default(ImmutableArray<string>));
```
I would expect this to throw, we shouldn't accept type symbols from other languages.

AlekseyTs replied:
I see, vbType is not a VB symbol, you are getting it from C# compilation.

jcouv replied:
Now I understand. I modified the test and confirmed we get `EnsureCSharpSymbolOrNull` to throw on VB symbols.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -2789,6 +2789,48 @@ protected override IPointerTypeSymbol CommonCreatePointerTypeSymbol(ITypeSymbol
+                    throw new ArgumentNullException($"{nameof(elementTypes)}[{i}]");
+                }
+
+                typesBuilder.Add(elementTypes[i].EnsureCSharpSymbolOrNull<ITypeSymbol, TypeSymbol>(nameof(elementTypes)));
```
I think we want this to throw for non-C# symbols, we cannot have null types for elements.

AlekseyTs commented:
LGTM

jaredpar commented:
:+1:

jcouv commented:
The last commit merges this change with feature flags which was implemented in parallel.

@CyrusNajmabadi Could you take a look at this last commit. I had to plumb the parse options through some editor feature tests and test utilities.

@AlekseyTs This last commit makes a small bug fix to `CSharpParseOptions`.

AlekseyTs commented on src/Compilers/CSharp/Portable/CSharpParseOptions.cs:
```diff
@@ -86,6 +86,7 @@ public override IEnumerable<string> PreprocessorSymbolNames
             kind: other.Kind,
             preprocessorSymbols: other.PreprocessorSymbols)
         {
+            _features = other._features;
```
It feels like we should simply pass ```other._features``` as the last argument in the constructor initializer.

jcouv replied:
Done.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/CSharpParseOptions.cs:
```diff
@@ -86,6 +86,7 @@ public override IEnumerable<string> PreprocessorSymbolNames
             kind: other.Kind,
             preprocessorSymbols: other.PreprocessorSymbols)
```
I'm not certain i even see the point of this constructor.  Parse options are immutable.  Why do we need a parse options constructor that just copies the values out of another parse options?

AlekseyTs replied:
The point is to clone and then adjust what is needed.

AlekseyTs replied:
Adjustments are done outside of this constructor

jcouv replied:
It is used by "With" methods that clone and modify one specific part with construction initializer.
For instance, `return new CSharpParseOptions(this) { Kind = kind };`

CyrusNajmabadi commented on src/EditorFeatures/Test/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs:
```diff
@@ -193,18 +194,19 @@ public abstract class AbstractCodeActionOrUserDiagnosticTest
             int index, IList<CodeAction> actions,
             IList<TextSpan> conflictSpans, IList<TextSpan> renameSpans, IList<TextSpan> warningSpans,
-            bool compareTokens)
+            bool compareTokens, ParseOptions parseOptions = null)
```
wrap.

jcouv replied:
Done. I moved parseOptions to a new line.

AlekseyTs commented:
LGTM
